lib/charms/vault_k8s/v0/vault_client.py

```diff
@@ -416,3 +447,100 @@ def get_num_raft_peers(self) -> int:
         """Return the number of raft peers."""
         raft_config = self._client.sys.read_raft_config()
         return len(raft_config["data"]["config"]["servers"])
+
+    def _get_autounseal_policy_name(self, relation_id: int) -> str:
+        """Return the policy name for the given relation id."""
+        return f"charm-autounseal-{relation_id}"
+
+    def _get_autounseal_approle_name(self, relation_id: int) -> str:
+        """Return the approle name for the given relation id."""
+        return f"charm-autounseal-{relation_id}"
+
+    def _configure_autounseal_policy(self, mount: str, relation_id: int, key_name: str) -> str:
+        """Create/update a policy within vault to use the transit key."""
+        mount_policy = textwrap.dedent(f"""
+        path "{mount}/encrypt/{key_name}" {{
+            capabilities = ["update"]
+        }}
+
+        path "{mount}/decrypt/{key_name}" {{
+            capabilities = ["update"]
+        }}
+        """)
+        policy_name = self._get_autounseal_policy_name(relation_id)
+        self._client.sys.create_or_update_policy(
+            policy_name,
+            mount_policy,
+        )
+        return policy_name
+
+    def _get_autounseal_key_name(self, relation_id: int) -> str:
+        """Return the key name for the given relation id."""
+        return str(relation_id)
```

Comment:
The idea here is to separate a design decision (the key name) from the rest of the logic. The reason I see this as valuable is because it separates the semantics of the two. The `relation_id` is _not_ the same thing as the key name. We have  decided to name them the same thing for now, but if this changes in the future, we will have one place to modify it, and don't need to go searching for instances of `relation_id` and make sure we replace all the correct instances.

I think this is good practice, but if others don't see the value I will remove.

Reply:
To emphasize this, look at [`_configure_autounseal_policy` ](https://github.com/canonical/vault-k8s-operator/blob/017c5793763c874060d80a05410a0849d504ce2d/lib/charms/vault_k8s/v0/vault_client.py#L459-L475)-- We accept both `relation_id` and `key_name` as parameters, even though they are the same (currently). This is because they are semantically different. I could accept just `relation_id`, but then I'd be embedding it in the policy where the `key_name` is, and if the key name ever gets changed the refactor just got a whole lot more complex, searching for all the little cases like this.

Again, I _think_ this is good practice and something I always actually try to make a conscious attempt to do, but I'd love to hear other perspectives.

Reply:
[If we choose to keep it] I will add a small note to this effect in the docstring.

Reply:
Makes sense. Unfortunately it needs to be here as this is where the vault interaction is that relies on it (unless we create a cross library dependency the other way, too).

Charm libraries are definitely pushing us to some less-than-ideal approaches. That being said, perhaps the outcome of https://warthogs.atlassian.net/browse/TLSENG-269 will improve this a bit.